global_helpers/panther_base_helpers.py

```diff
@@ -262,9 +262,12 @@ def deep_get(dictionary: dict, *keys, default=None):
 
     Inspired by https://bit.ly/3a0hq9E
     """
-    return reduce(
+    out = reduce(
         lambda d, key: d.get(key, default) if isinstance(d, Mapping) else default, keys, dictionary
     )
+    if out is None:
```

Comment:
If `default=None` (the default parameter value) `out` will also be `None` if the key is not found. So for a call that was:

```python
field_value = deep_get(event, "field")
# None check means the field isn't there or has a `null` JSON value
if field_value is None:
  # do something 
```

This seems as potentially breaking behavior and since this is a core helper function that's been defined since the early days, I have some suggestions to help in case this is an existing use case (best is first in my opinion):
1. Introduce a boolean flag to allow callers to selectively enable this behavior: `deep_get(dictionary, *keys, default=None, use_default_on_none=False)`
2. Introduce a boolean flag to allow someone to disable this behavior: `deep_get(dictionary, *keys, default=None, use_default_on_none=True)`
3. Introduce a new function that wraps the `deep_get` call with the new behavior.

Reply:
If I'm understanding your concern correctly, the historical behavior is unchanged.

historically the behavior was 
```python
event = {}
field_value = deep_get(event, "does_not_exist")
# Old behavior is field_value is None
# The below handling will continue to be correct without change
if field_value is None:
   # do something
```
I added this unit test to a copy of panther-analysis that does not have the update to deep_get that is in the PR. 
```python
class TestOldDeepGet(unittest.TestCase):
    def test_f_d(self):
        event = {}
        self.assertEqual(p_b_h.deep_get(event, 'does_not_exist'), None)
```

I added a test case for the concern above to this implementation of deep_get to the global helpers unit test suite [here](https://github.com/panther-labs/panther-analysis/pull/672/files#diff-ba7b45aeb238fff6523363f293e24aaf8e9ec36b328f86da401bfbde91b93148R462)

Reply:
You're right, sorry for the incorrect example and the misunderstanding. It is true that `deep_get(event, "does_not_exist")` behavior is unmodified. What changes is that whether someone relied on the `None` return value to check for null values:

```python
event = {"request": {"does_not_exist": None}}
# Previous behavior
if deep_get(event, 'request', 'does_not_exist', '') is None: # True
# New behavior
if deep_get(event, 'request', 'does_not_exist', '') is None: # False
```

Just noting this down for my own understanding mostly 😅